### PR TITLE
Using Arrow Keys to Move in Panorama

### PR DESCRIPTION
**Pull request by danZzyy**
Even though this is a built in feature in GSV, it was not working in ours for some reason. Maybe because of the other layers interfering with the focus. 
This is not quite done. I am looking for feedback here. Sometimes the links on the SV (the arrows on the ground) do not respond. When they do respond you go  in the correct direction. I originally thought my math was wrong but I am not sure now.

### CODE REVIEW

**Comment by danZzyy** [comment]
One thing I've just discovered is that the panorama's pov heading is 0-360 while the link headings are -180-180

**Comment by kotarohara** [comment]
I'll review this when I get a chance. Probably tomorrow. I haven't seen the code yet, but:

1. What do you mean by the arrow does not respond? If you are getting a panorama id, you can just set the panorama id by svl.panorama.setPano(), no? (which you seem to do)
2. They are same if do: heading + 360 (mod 360).

**Comment by danZzyy** [comment]
Kotaro, I mean based on the orientation on the arrows relative to the pov (like if you set a link perpendicular to a direction) if you press a direction that very obviously corresponds to a link, it sometimes still does not go to it.

**Comment by danZzyy** [comment]
(Look at the second to last commit, the last one was undoing the damage of pulling which gave me duplicate methods with the old code)
I improved variable names, disabled scrolling as a result of arrow keys because that was firing too, and included a condition that fixed an issue I had before when if only one link exists, it was not recognized.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 1]
```diff
@@ -54,6 +97,24 @@ function Keyboard ($) {
      * @private
      */
     function documentKeyUp (e) {
+        switch (e.keyCode) {
+            // "left"
+            case 37:
+                movePano(270);
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Should left and right keys make people step on sides or change the heading angle? The default GSV API changes the heading angle. What is your rational for stepping left or right?

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 2]
```diff
@@ -54,6 +97,24 @@ function Keyboard ($) {
...
+                break;
+            // "up"
+            case 38:
+                movePano(0);
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: movePano is not a very descriptive function name... Change it to moveForward and moveBackward?

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 3]
```diff
@@ -1,10 +1,53 @@
...
+        diff=Math.abs(360-diff);
+
+    return diff;
+
+}
+
+function movePano(angle) {
+  var curr;
+  var found = false;
+  for(var i = 0; i < svl.panorama.links.length; i++) {
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: define a variable `len` to store svl.panorama.links.length before the loop.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 4]
```diff
@@ -1,10 +1,53 @@
+window.addEventListener("keydown", function(e) {
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Why do you need a new event listener? Look at line 73 and 74. We listeners are already attached to `document`.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 5]
```diff
@@ -1,10 +1,53 @@
...
+    // space and arrow keys
+    if([32, 37, 38, 39, 40].indexOf(e.keyCode) > -1) {
+        e.preventDefault();
+    }
+}, false);
+
+ // taken from 
+ // http://stackoverflow.com/questions/7799722/how-do-i-implement-move-forward-action-for-google-street-view
+ function difference(link, angle) {
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: 1. I don't think you need to have a separate function for this... You could, but it looks very simple...
2. The function name is not very meaningful. What do you mean by difference?
3. I prefer not to have functions in global scope. Put both difference() and movePano in the Keyboard module so they are private.

**Comment by kotarohara** [comment]
I added comments to your code. Please update and let me know once you are done.

**Comment by danZzyy** [comment]
I hadn't realized that left an right control rotation and not movement. Makes a lot of sense that you would only be moving in the direction you are facing.

**Comment by kotarohara** [comment]
I have just tried actual Street View and left/right corresponds to updating heading angles. Let's abide by what they do :) 

Also, I've skimmed your code again. Wouldn't it simpler if you take difference in heading and link angles, take cosine, and see if the value is larger than diff=x-degrees? That way you can get rid of weird if statement...

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 6]
```diff
@@ -30,13 +32,76 @@ function Keyboard ($) {
...
     function documentKeyDown(e) {
+        if([32, 37, 38, 39, 40].indexOf(e.keyCode) > -1) { 
+            e.preventDefault();
+        }
+    
+        switch (e.keyCode) {
+            // "left"
+            // rotate pano
+            case 37:
+                rotatePov(-2);
```
Comment (danZzyy) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Originally tried with +/-1 which was fine but a little slow. The rate of the rotation depends on the browser's refresh rate.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 7]
```diff
@@ -30,13 +30,62 @@ function Keyboard ($) {
...
     }
+    function toRadians (angle) {
+        return angle * (Math.PI / 180);
+    }
+    // Move in the direction of a link closest to a given angle
+    function movePano(angle) {
+        // take the cosine of the difference for each link to the current heading in radians and stores them to an array
+        var cosines = svl.panorama.links.map(function(link) { return Math.cos(toRadians(svl.panorama.pov.heading + angle) - toRadians(link.heading))});
+        // finds index of greatest value in cosines array
+        var maxIndex = cosines.reduce((iMax, x, i, arr) => x > arr[iMax] ? i : iMax, 0);
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Unfortunately `=>` operator is not supported on Safari and IE :( You have to use an old anonymous function.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 8]
```diff
@@ -30,13 +30,62 @@ function Keyboard ($) {
...
+        //in the case of one link, this prevents you from moving in that direction if you press the opposite
+        if(cosines[maxIndex] > 0.5){
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Oh, so you decided to use 45-degrees? I thought you were going to use 90 degrees as a threshold. Either way is fine. Let's try.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 9]
```diff
@@ -30,13 +30,62 @@ function Keyboard ($) {
         $(document).bind('keyup', documentKeyUp);
         $(document).bind('keydown', documentKeyDown);
     }
+    function toRadians (angle) {
+        return angle * (Math.PI / 180);
+    }
+    // Move in the direction of a link closest to a given angle
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Small thing, but I prefer a comment before a method name to be blocked by /** */ . JSDoc automatically parses this and generates documentation this way.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 10]
```diff
@@ -30,13 +30,62 @@ function Keyboard ($) {
...
+        if(cosines[maxIndex] > 0.5){
+            // transitions panorama to the link of the greatest cosine (closest to direction)
+            svl.panorama.setPano(svl.panorama.links[maxIndex].pano);
+        }
+    }
+    // abstract movePano with more meaningful function name
+    function moveForward(){
+        movePano(0);
+    }
+    function moveBackward(){
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: You should have a line break.

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 11]
```diff
@@ -30,13 +30,62 @@ function Keyboard ($) {
...
+        movePano(180);
+    }
+
+    function rotatePov(degree){
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Comment

File: public/javascripts/SVLabel/src/SVLabel/Keyboard.js [hunk 12]
```diff
@@ -30,13 +30,62 @@ function Keyboard ($) {
...
+        svl.panorama.setPov({heading: heading, pitch: pitch, zoom: zoom});
+    }
 
     /**
      * This is a callback for a key down event
      * @param {object} e An event object
      * @private
      */
     function documentKeyDown(e) {
+        // lock scrolling in response to key pressing
```
Comment (kotarohara) [inline public/javascripts/SVLabel/src/SVLabel/Keyboard.js]: Nice comment 👍  I was wondering why you needed to prevent default.

**Comment by danZzyy** [comment]
rather than using the reduce with an anonymous function, I just used Math.max and indexOf since those make more sense to me. The map reduce line I had was taken from stackoverflow originally.

**Comment by kotarohara** [comment]
Ok, I'm merging this and closing #105

**Pull request merged**